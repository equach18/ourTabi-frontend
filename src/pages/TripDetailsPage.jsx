import { useContext, useEffect } from "react";
import TripContext from "../context/TripContext";
import { Link, useNavigate } from "react-router-dom";
import MemberPanel from "../components/member/MemberPanel";
import CommentPanel from "../components/comment/CommentPanel";
import ActivityList from "../components/activity/ActivityList";
import TripDetails from "../components/trip/TripDetails";

function TripDetailsPage({ currentUser, removeTrip }) {
  const { trip, activities, members } = useContext(TripContext);
  const navigate = useNavigate();

  useEffect(() => {
    if (!trip) {
      navigate("/not-found");
    }
  }, [trip, navigate]);

  if (!trip || !currentUser) return null;

  const isMember = members?.some((m) => m.userId === currentUser?.id);
  const isTripCreator = trip?.creatorId === currentUser?.id;

  async function handleDelete() {
    if (window.confirm("Are you sure you want to delete this trip?")) {
      await removeTrip(trip.id);
      navigate("/dashboard");
    }
  }

  return (
    <div className="h-[calc(100vh-4rem)] grid grid-cols-1 lg:grid-cols-[280px_1fr_280px] gap-6 px-6 pt-10 pb-12 bg-gradient-to-r from-stone-300 via-white to-stone-300">
      <div className="flex flex-col gap-6 overflow-hidden">
        {/* Trip Info */}
        <div className="bg-white p-4 rounded-xl shadow max-h-[45%] overflow-y-auto">
          <h1 className="text-2xl font-bold mb-2">{trip.title}</h1>
          <TripDetails trip={trip} />
          {isTripCreator && (
            <div className="space-y-2 mt-4">
              <Link
                to={`/trips/${trip.id}/edit`}
                className="block w-full bg-blue-500 text-white py-1.5 rounded-md text-sm text-center hover:bg-blue-600 transition"
              >
                Edit Trip
              </Link>
              <button
                onClick={handleDelete}
                className="w-full bg-gray-500 text-white py-1.5 rounded-md text-sm hover:bg-gray-600 transition"
              >
                Delete Trip
              </button>
            </div>
          )}
        </div>

        {/* Member Panel */}
        {isMember && (
          <div className="bg-white rounded-xl shadow flex-1 overflow-y-auto">
            <MemberPanel isTripCreator={isTripCreator} members={members} />
          </div>
        )}
      </div>

      {/* Activities */}
      <div className="bg-white p-4 rounded-xl shadow h-full overflow-y-auto">
        <ActivityList activities={activities} trip={trip} />
      </div>

      {/* Comments */}
      {isMember && (
        <div className="bg-white rounded-xl shadow flex flex-col h-full overflow-hidden">
          <CommentPanel />
        </div>
      )}
    </div>
  );
}

export default TripDetailsPage;
