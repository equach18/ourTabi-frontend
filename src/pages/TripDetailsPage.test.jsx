vi.mock("../components/trip/TripDetails", () => ({
  default: ({ trip }) => (
    <div data-testid="TripDetails">Trip Details - {trip?.title}</div>
  ),
}));
vi.mock("../components/member/MemberPanel", () => ({
  default: () => <div>Member Panel</div>,
}));
vi.mock("../components/comment/CommentPanel", () => ({
  default: () => <div>Comment Panel</div>,
}));
vi.mock("../components/activity/ActivityList", () => ({
  default: () => <div>Activity Section</div>,
}));

import { screen } from "@testing-library/react";
import TripDetailsPage from "./TripDetailsPage";
import { Routes, Route } from "react-router-dom";
import {
  renderWithAllProviders,
  testTrip,
  testUser1,
  defaultMembers,
  defaultTripContext,
} from "../test-helpers/_testCommons";

describe("TripDetailsPage", () => {
  function renderDetailsPage(overrides = {}) {
    const {
      currentUser = testUser1,
      trip = testTrip,
      activities = [],
      members = defaultMembers,
      route = `/trips/${testTrip.id}`,
      userContext = { currentUser, friends: [] },
      tripContext: customTripContext = {},
    } = overrides;

    const tripContext = {
      ...defaultTripContext,
      trip,
      activities,
      members,
      ...customTripContext,
    };

    return renderWithAllProviders(
      <Routes>
        <Route
          path="/trips/:tripId"
          element={
            <TripDetailsPage currentUser={currentUser} removeTrip={vi.fn()} />
          }
        />
        <Route path="*" element={<div>not-found</div>} />
      </Routes>,
      {
        route,
        userContext,
        tripContext,
      }
    );
  }
  it("matches snapshot", () => {
    const { asFragment } = renderDetailsPage();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders trip details and panels for member user", () => {
    const currentUser = { id: 1, username: "testuser" };

    renderDetailsPage({
      currentUser,
      trip: testTrip,
      members: [{ userId: 1 }],
    });

    expect(screen.getByTestId("TripDetails")).toHaveTextContent(
      `Trip Details - ${testTrip.title}`
    );
    expect(screen.getByText(/Activity/i)).toBeInTheDocument();
    expect(screen.getByText(/Comment/i)).toBeInTheDocument();
    expect(screen.getByText(/Member/i)).toBeInTheDocument();
  });
  it("does not show edit/delete buttons if user is not creator", () => {
    renderDetailsPage({
      currentUser: { ...testUser1, id: 345 },
    });

    expect(screen.queryByText(/Edit Trip/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/Delete Trip/i)).not.toBeInTheDocument();
  });

  it("does not show member or comment panel if user is not a trip member", () => {
    renderDetailsPage({
      currentUser: { ...testUser1, id: 999 },
      members: [],
    });

    expect(screen.queryByText(/Comment Panel/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/Member Panel/i)).not.toBeInTheDocument();
  });

  it("navigates to not-found if trip is missing", () => {
    renderDetailsPage({
      trip: null,
      route: "/trips/999",
    });

    expect(true).toBe(true);
  });
});
