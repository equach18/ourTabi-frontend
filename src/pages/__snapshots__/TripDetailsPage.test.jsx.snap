// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TripDetailsPage > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-[calc(100vh-4rem)] grid grid-cols-1 lg:grid-cols-[280px_1fr_280px] gap-6 px-6 pt-10 pb-12 bg-gradient-to-r from-stone-300 via-white to-stone-300"
  >
    <div
      class="flex flex-col gap-6 overflow-hidden"
    >
      <div
        class="bg-white p-4 rounded-xl shadow max-h-[45%] overflow-y-auto"
      >
        <h1
          class="text-2xl font-bold mb-2"
        >
          Test trip
        </h1>
        <div
          data-testid="TripDetails"
        >
          Trip Details - Test trip
        </div>
        <div
          class="space-y-2 mt-4"
        >
          <a
            class="block w-full bg-blue-500 text-white py-1.5 rounded-md text-sm text-center hover:bg-blue-600 transition"
            data-discover="true"
            href="/trips/42/edit"
          >
            Edit Trip
          </a>
          <button
            class="w-full bg-gray-500 text-white py-1.5 rounded-md text-sm hover:bg-gray-600 transition"
          >
            Delete Trip
          </button>
        </div>
      </div>
      <div
        class="bg-white rounded-xl shadow flex-1 overflow-y-auto"
      >
        <div>
          Member Panel
        </div>
      </div>
    </div>
    <div
      class="bg-white p-4 rounded-xl shadow h-full overflow-y-auto"
    >
      <div>
        Activity Section
      </div>
    </div>
    <div
      class="bg-white rounded-xl shadow flex flex-col h-full overflow-hidden"
    >
      <div>
        Comment Panel
      </div>
    </div>
  </div>
</DocumentFragment>
`;
